File: Leetcode_solutions/Symmetric Tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isSymmetric = function(root) {
  if (!root) {
    // Sanity check
    return true;
  }

  function isMirror(s, t) {
    var q1 = [s],
      q2 = [t];

    // Perform breadth-first search
    while (q1.length > 0 || q2.length > 0) {
      // Dequeue
      var n1 = q1.shift(),
        n2 = q2.shift();

      // Two null nodes, let's continue
      if (!n1 && !n2) continue;

      // Return false as long as there is a mismatch
      if (!n1 || !n2 || n1.val !== n2.val) return false;

      // Scan tree s from left to right
      // and scan tree t from right to left
      q1.push(n1.left);
      q1.push(n1.right);
      q2.push(n2.right);
      q2.push(n2.left);
    }

    return true;
  }
  return isMirror(root.left, root.right);
};

console.log(isSymmetric([1, 2, 2, 3, 4, 4, 3]));

File: Leetcode_solutions/Symmetric_Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution:
    def isSymmetric(self, root: TreeNode) -> bool:
        return self.isMirror(root, root)

    def isMirror(self, t1, t2):
        if t1 == None and t2 == None:
            return True
        if t1 == None or t2 == None:
            return False
            return t1.val == t2.val and self.isMirror(t1.right, t2.left) and self.isMirror(t1.left, t2.right)


class Solution:
    def isSymmetric(self, root: TreeNode) -> bool:
        q = []
        q.append(root)
        q.append(root)
        while len(q) > 0:
            t1 = q.pop()
            t2 = q.pop()
            if t1 == None and t2 == None:
                continue
            if t1 == None or t2 == None:
                return False
            if t1.val != t2.val:
                return False
            q.append(t1.left)
            q.append(t2.right)
            q.append(t1.right)
            q.append(t2.left)
        return True
